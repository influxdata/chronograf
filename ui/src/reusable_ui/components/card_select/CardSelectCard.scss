/*
  Card Selector Card
  ------------------------------------------------------------------------------
*/

$card-select--gutter: 4px;

$card-breakpoint-col-2: 600px;
$card-breakpoint-col-3: 700px;
$card-breakpoint-col-4: 800px;
$card-breakpoint-col-5: 900px;
$card-breakpoint-col-6: 1000px;

.card-select--card-holder {
    width: 100%;
    padding-bottom: 100%;
    float: left;
    position: relative;

    @media only screen and (min-width: $card-breakpoint-col-2) {
        width: 50%;
        padding-bottom: 50%;
    }

    @media only screen and (min-width: $card-breakpoint-col-3) {
        width: 33.3333%;
        padding-bottom: 33.3333%;
    }

    @media only screen and (min-width: $card-breakpoint-col-4) {
        width: 25%;
        padding-bottom: 25%;
    }

    @media only screen and (min-width: $card-breakpoint-col-5) {
        width: 20%;
        padding-bottom: 20%;
    }

    @media only screen and (min-width: $card-breakpoint-col-6) {
        width: 16.6667%;
        padding-bottom: 16.6667%;
    }
}

.card-select--card {
    background-color: $g1-raven;
    border-radius: $radius;
    border: $ix-border solid transparent;
    width: calc(100% - #{$card-select--gutter});
    height: calc(100% - #{$card-select--gutter});
    margin: $card-select--gutter / 2;
    position: absolute;
    transition: border-color .2s ease, background-color .2s ease;
    box-sizing: border-box;
}

.card-select--container {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    cursor: pointer;
    font-size: 22px;
    user-select: none;
    width: 100%;
    height: 100%;
    display: inline-flex;
    justify-content: center;
    flex-direction: column;
    align-items: center;
    input {
        position: absolute;
        opacity: 0;
        cursor: pointer;
    }
}

.card-select--image {
    justify-content: center;
    width: 60%;
    .card-select--placeholder {
        color: $ix-text-default;
        transition: color .2s ease;
    }
    img {
        filter: grayscale(0.80);
        width: 100%;
        height: 100%;
        transition: filter .2s ease;
        font-size: $ix-text-base;
    }
}

.card-select--active:hover,
.card-select--active.card-select--checked {
    background-color: $g5-pepper;
    .card-select--image {
        img {
            filter: grayscale(0);
        }
    }
    .card-select--placeholder {
        color: $c-pool;
    }
}

.card-select--active.card-select--checked {
    border-color: $c-rainforest;
}

.card-select--disabled {
    opacity: .3;
    .card-select--container {
        cursor: not-allowed;
    }
}

.card-select--label {
  text-align: center;
  min-width: 100%;
  display: inline-block;
  position: absolute;
  bottom: 5%;
  font-family: $ix-text-font;
  color: $ix-text-default;
  font-size: $ix-text-base;
  font-weight: 600;
}

.card-select--checkmark {
    position: absolute;
    left: 5%;
    top: 5%;
    height: 22px;
    width: 22px;
    color: $c-rainforest;
    opacity: 0;
    transition: opacity .2s ease;
}

.card-select--checkmark.card-select--checked {
    opacity: 1;
}
