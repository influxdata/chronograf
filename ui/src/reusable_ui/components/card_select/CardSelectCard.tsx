// Libraries
import React, {PureComponent} from 'react'
import classnames from 'classnames'
import {ErrorHandling} from 'src/shared/decorators/errors'
import ProtoboardIcon from 'src/reusable_ui/components/card_select/ProtoboardIcon'

interface Props {
  id: string
  name?: string
  label: string
  image?: string
  checked?: boolean
  disabled?: boolean
  onClick: () => void
}

@ErrorHandling
class CardSelectCard extends PureComponent<Props> {
  public static defaultProps: Partial<Props> = {
    checked: false,
    disabled: false,
  }

  public render() {
    const {id, label, checked, name, disabled} = this.props

    return (
      <div
        data-toggle="card_toggle"
        onClick={this.handleClick}
        className={classnames('card-select--card', {
          'card-select--checked': checked,
          'card-select--disabled': disabled,
          'card-select--active': !disabled,
        })}
      >
        <label className="card-select--container">
          <input
            id={`card_select_${id}`}
            name={name}
            type="checkbox"
            value={id}
            defaultChecked={checked}
            disabled={disabled}
          />
          <span
            className={classnames(
              'card-select--checkmark',
              'icon',
              'checkmark',
              {
                'card-select--checked': checked,
              }
            )}
          />
          <div className="card-select--image">{this.cardImage}</div>
          <span className="card-select--label">{label}</span>
        </label>
      </div>
    )
  }

  private get cardImage() {
    const {image, label} = this.props

    if (image) {
      return <img src={image} alt={`${label} icon`} />
    }

    return <ProtoboardIcon displayText={label} />
  }

  private handleClick = e => {
    const {onClick, disabled} = this.props
    e.preventDefault()
    if (!disabled) {
      onClick()
    }
  }
}

export default CardSelectCard
