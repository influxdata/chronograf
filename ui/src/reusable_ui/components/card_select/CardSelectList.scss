.card-select--wrapper {
  width: 100%;
  border: none;
  box-sizing: border-box;
  legend {
    font-weight: 600;
  }
}

.card-select--cards {
  display: inline-flex;
  flex-wrap: wrap;
}