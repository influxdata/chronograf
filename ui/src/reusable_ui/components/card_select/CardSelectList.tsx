import React, {PureComponent, ReactElement} from 'react'
import {ErrorHandling} from 'src/shared/decorators/errors'

import {CardSelectCardProps} from 'src/types/cardSelect'

import 'src/reusable_ui/components/card_select/CardSelectList.scss'

interface Props {
  children: Array<ReactElement<CardSelectCardProps>>
  legend: string
}

@ErrorHandling
class CardSelectList extends PureComponent<Props> {
  public render() {
    const {children} = this.props

    return (
      <fieldset className="card-select--wrapper">
        <div className="card-select--cards">{children}</div>
      </fieldset>
    )
  }
}

export default CardSelectList
