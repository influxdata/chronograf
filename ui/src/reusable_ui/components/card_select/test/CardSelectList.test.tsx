import React from 'react'
import {shallow} from 'enzyme'

import CardSelectList from 'src/reusable_ui/components/card_select/CardSelectList'
import CardSelectCard from 'src/reusable_ui/components/card_select/CardSelectCard'

describe('Card Select Card', () => {
  let wrapper

  const wrapperSetup = (override = {}) => {
    const props = {
      children: undefined,
      legend: 'legend',
      ...override,
    }

    return shallow(<CardSelectList {...props} />)
  }

  const childSetup = (override = {}) => {
    const props = {
      id: 'card_id',
      label: 'Card Label',
      image: undefined,
      checked: undefined,
      disabled: undefined,
      ...override,
    }

    return shallow(<CardSelectCard {...props} />)
  }

  const cardChildren = [childSetup(), childSetup()]

  beforeEach(() => {
    wrapper = wrapperSetup({children: cardChildren})
  })

  it('mounts without exploding', () => {
    expect(wrapper).toHaveLength(1)
  })

  it('renders one fieldset', () => {
    expect(wrapper.find('fieldset')).toHaveLength(1)
  })

  it('renders one legend', () => {
    expect(wrapper.find('legend')).toHaveLength(1)
  })

  it('matches snapshot with two children', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
