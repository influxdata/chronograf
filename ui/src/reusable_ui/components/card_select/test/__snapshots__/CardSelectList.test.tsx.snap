// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Select Card matches snapshot with two children 1`] = `
<fieldset
  className="card-select--wrapper"
>
  <legend>
    legend
  </legend>
  <div
    className="card-select--cards"
  >
    <div
      className="card-select--card card-select--active"
      data-toggle="card_toggle"
      onClick={[Function]}
    >
      <label
        className="card-select--container"
      >
        <span
          className="card-select--label"
        >
          Card Label
        </span>
        <input
          checked={false}
          disabled={false}
          id="card_select_card_id"
          name="card_select_card_id"
          type="checkbox"
          value="card_id"
        />
        <span
          className="card-select--checkmark icon checkmark"
        />
        <div
          className="card-select--image"
        >
          <span
            className="card-select--placeholder icon dash-j"
          />
        </div>
      </label>
    </div>
    <div
      className="card-select--card card-select--active"
      data-toggle="card_toggle"
      onClick={[Function]}
    >
      <label
        className="card-select--container"
      >
        <span
          className="card-select--label"
        >
          Card Label
        </span>
        <input
          checked={false}
          disabled={false}
          id="card_select_card_id"
          name="card_select_card_id"
          type="checkbox"
          value="card_id"
        />
        <span
          className="card-select--checkmark icon checkmark"
        />
        <div
          className="card-select--image"
        >
          <span
            className="card-select--placeholder icon dash-j"
          />
        </div>
      </label>
    </div>
  </div>
</fieldset>
`;
